infrastructure/ssz/src/main/java/tech/pegasys/teku/infrastructure/ssz/tree/MerkleUtil.java

```diff
@@ -0,0 +1,114 @@
+/*
+ * Copyright ConsenSys Software Inc., 2022
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package tech.pegasys.teku.infrastructure.ssz.tree;
+
+import com.google.common.annotations.VisibleForTesting;
+import java.util.ArrayDeque;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Queue;
+import org.apache.commons.lang3.tuple.Pair;
+import org.apache.tuweni.bytes.Bytes32;
+
+public class MerkleUtil {
+
+  /**
+   * Returns the merkle inclusion proof for the field at the given leaf index in the given tree. See
+   * https://github.com/ethereum/consensus-specs/blob/dev/ssz/merkle-proofs.md#merkle-multiproofs
+   * for more info on merkle proofs.
+   */
+  public static List<Bytes32> constructMerkleProof(TreeNode root, long leafGeneralizedIndex) {
+    if (leafGeneralizedIndex <= 1) {
+      // Nothing to prove or invalid index.
+      return Collections.emptyList();
+    }
+    // Path represented by generalized indices.
+    List<Long> rootToLeafPath = getPathToNode(leafGeneralizedIndex);
+
+    // The sibling of every node along the path is part of the proof.
+    List<Bytes32> proof = new ArrayList<>();
+
+    for (int i = 0; i < rootToLeafPath.size() - 1; i++) {
+      long currentNodeIndex = rootToLeafPath.get(i);
+      long nextNodeIndex = rootToLeafPath.get(i + 1);
+
+      if (GIndexUtil.gIdxLeftGIndex(currentNodeIndex) == nextNodeIndex) {
+        // Next node is left, so the right node is in the proof.
+        proof.add(root.get(GIndexUtil.gIdxRightGIndex(currentNodeIndex)).hashTreeRoot());
+      } else if (GIndexUtil.gIdxRightGIndex(currentNodeIndex) == nextNodeIndex) {
+        // Next node is right, so the left node is in the proof.
+        proof.add(root.get(GIndexUtil.gIdxLeftGIndex(currentNodeIndex)).hashTreeRoot());
+      } else {
+        throw new IllegalArgumentException("Invalid path from root to leaf: " + rootToLeafPath);
+      }
+    }
+
+    Collections.reverse(proof);
+    return Collections.unmodifiableList(proof);
+  }
+
+  /**
+   * Returns the path from the root to the specified node as an array of generalized indices.
+   * path[0] is the root itself (gIndex of 1) path[path.length - 1] is the node (gIndex of
+   * `nodeIndex`)
+   */
+  @VisibleForTesting
+  static List<Long> getPathToNode(long nodeIndex) {
```

Comment:
nit: do these map to functions in the spec? The tests are good, just wondering if there's an algorithm I'm checking them against, and if we can make that relationship a little clearer if they're just a spec function by naming them in a way we can relate them directly...

Reply:
The spec doesn't define these functions from what I can see. The light client spec defines a [`compute_merkle_proof_for_state`](https://github.com/ethereum/consensus-specs/blob/dev/specs/altair/light-client/full-node.md#compute_merkle_proof_for_state) whose implementation is not provided. The impl in this PR is more general purpose than just an interface accepting `BeaconState`. The [`Merkle multiproofs`](https://github.com/ethereum/consensus-specs/blob/dev/ssz/merkle-proofs.md#merkle-multiproofs) spec does not provide these functions either. The only definitions I could see came from the test cases (linked in OP).

Reply:
ok all good, just makes it hard to verify against a spec :)